laws/shared/src/test/scala/cats/effect/MemoizeTests.scala

```diff
@@ -92,4 +92,115 @@ class MemoizeTests extends BaseTestsSuite {
       Concurrent.memoize(fa).flatten <-> fa
     }
   }
+
+  testAsync("Memoized effects can be canceled when there are no other active subscribers (1)") { implicit ec =>
+    implicit val cs = ec.contextShift[IO]
+    implicit val timer = ec.timer[IO]
+
+    val prog = for {
+      completed <- Ref[IO].of(false)
+      action = IO.sleep(200.millis) >> completed.set(true)
+      memoized <- Concurrent.memoize(action)
+      fiber <- memoized.start
+      _ <- IO.sleep(100.millis)
+      _ <- fiber.cancel
+      _ <- IO.sleep(300.millis)
+      res <- completed.get
+    } yield res
+
+    val result = prog.unsafeToFuture()
+    ec.tick(500.millis)
+    result.value shouldBe Some(Success(false))
+  }
+
+  testAsync("Memoized effects can be canceled when there are no other active subscribers (2)") { implicit ec =>
+    implicit val cs = ec.contextShift[IO]
+    implicit val timer = ec.timer[IO]
+
+    val prog = for {
+      completed <- Ref[IO].of(false)
+      action = IO.sleep(300.millis) >> completed.set(true)
+      memoized <- Concurrent.memoize(action)
+      fiber1 <- memoized.start
+      _ <- IO.sleep(100.millis)
+      fiber2 <- memoized.start
+      _ <- IO.sleep(100.millis)
+      _ <- fiber2.cancel
+      _ <- fiber1.cancel
+      _ <- IO.sleep(400.millis)
+      res <- completed.get
+    } yield res
+
+    val result = prog.unsafeToFuture()
+    ec.tick(600.millis)
+    result.value shouldBe Some(Success(false))
+  }
+
+    testAsync("Memoized effects can be canceled when there are no other active subscribers (3)") { implicit ec =>
+    implicit val cs = ec.contextShift[IO]
+    implicit val timer = ec.timer[IO]
+
+    val prog = for {
+      completed <- Ref[IO].of(false)
+      action = IO.sleep(300.millis) >> completed.set(true)
+      memoized <- Concurrent.memoize(action)
+      fiber1 <- memoized.start
+      _ <- IO.sleep(100.millis)
+      fiber2 <- memoized.start
+      _ <- IO.sleep(100.millis)
+      _ <- fiber1.cancel
+      _ <- fiber2.cancel
+      _ <- IO.sleep(400.millis)
+      res <- completed.get
+    } yield res
+
+    val result = prog.unsafeToFuture()
+    ec.tick(600.millis)
+    result.value shouldBe Some(Success(false))
+  }
+
+  testAsync("Running a memoized effect after it was previously canceled reruns it") { implicit ec =>
+    implicit val cs = ec.contextShift[IO]
+    implicit val timer = ec.timer[IO]
+
+    val prog = for {
+      started <- Ref[IO].of(0)
+      completed <- Ref[IO].of(0)
+      action = started.update(_ + 1) >> timer.sleep(200.millis) >> completed.update(_ + 1)
+      memoized <- Concurrent.memoize(action)
+      fiber <- memoized.start
+      _ <- IO.sleep(100.millis)
+      _ <- fiber.cancel
+      _ <- memoized.timeout(1.second)
+      v1 <- started.get
+      v2 <- completed.get
+    } yield v1 -> v2
+
+    val result = prog.unsafeToFuture()
+    ec.tick(500.millis)
+    result.value shouldBe Some(Success(2 -> 1))
```

Comment:
Style nit: I prefer to use `->` for associations, not for something you just happen to have two of.

```suggestion
    result.value shouldBe Some(Success((2 , 1)))
```